Fix scaffold template with namespace and `--model-name` option

kenzo-tanaka opened the pull request:
close https://github.com/rspec/rspec-rails/issues/2532

Fixed the behavior of scaffold template when namespace and `--model-name` option passed.

```shell
$ rails g scaffold_controller admin/products --model-name=product
```

```rb
describe "POST /create" do
  context "with valid parameters" do
    it "creates a new Product" do
      expect {
        # Here, I expected `{ product: valid_attributes }` like the test of update action
        post admin_products_url, params: { admin_product: valid_attributes } 
      }.to change(Product, :count).by(1)
    end
    # ...
end
```

This is because `ns_file_name` method that is called in scaffold template returns `admin_product` when namespaced.

https://github.com/rspec/rspec-rails/blob/ba2d66abf275fe1f490c3ed219459e42698936d6/lib/generators/rspec/scaffold/scaffold_generator.rb#L79-L83

So, I replaced `ns_file_name` to `singular_table_name` in scaffold templates.

Also, I added some tests to ensure that templates works in the old way when the `--model-name` option is not passed.

pirj commented:
Nice, thank you!

What happens if you don't specify ` --model-name=card` in `rails g scaffold_controller admin/cards`, will it behave in the old way, will it use `admin_cards` (`admin_products` from the ticket's example)?

In case no specs fail, would you please add some examples to prevent from regression to `spec/generators/rspec/scaffold/scaffold_generator_spec.rb`?

kenzo-tanaka commented:
@pirj Thank you for quick response!

> What happens if you don't specify --model-name=card in rails g scaffold_controller admin/cards, will it behave in the old way, will it use admin_cards (admin_products from the ticket's example)?

Thanks for pointing that out. If I remove `--model-name=card`, it behaved in the old way like below:

```rb
describe "POST /create" do
  context "with valid parameters" do
    it "creates a new Admin::Card" do
      expect {
        post admin_cards_url, params: { admin_card: valid_attributes }
      }.to change(Admin::Card, :count).by(1)
    end
   # ...
  end
end
```

Is this the behavior that you expected?

---

> In case no specs fail, would you please add some examples to prevent from regression to spec/generators/rspec/scaffold/scaffold_generator_spec.rb?

OK! I'll try it.

pirj commented:
> Is this the behavior that you expected?

Yes, perfect. Thank you!

pirj reviewed (approved):
Perfect, thank you!

kenzo-tanaka commented:
@pirj 
When I run `rails g scaffold_controller admin/cards --model-name=card  --controller_specs` in a example app, the controller spec may have a same situation of this issue.

```rb
describe "POST #create" do
  context "with valid params" do
    it "creates a new Card" do
      expect {
        post :create, params: {admin_card: valid_attributes}, session: valid_session
      }.to change(Card, :count).by(1)
    end
    # ...
  end
end
```

Should this file be fixed as well?

https://github.com/rspec/rspec-rails/blob/d3e7b85877fcbcec63f8a76434d8750e7f3b7aef/lib/generators/rspec/scaffold/templates/controller_spec.rb#L82

pirj commented:
It would be great if you're up for it.

PedroAugustoRamalhoDuarte commented:
We have to change templates/api_request_spec.rb too, there is the same problem.

kenzo-tanaka commented:
@pirj @PedroAugustoRamalhoDuarte 
Finally, I fixed the following files because they had the same problem.

- `templates/request_spec.rb`
- `templates/api_request_spec.rb`
- `templates/controller_spec.rb`
- `templates/api_controller_spec.rb`

And also, I added some examples to prevent from regression. Please check🙏

pirj reviewed (approved):
Great job! Thanks a lot for taking care of the whole thing.

JonRowe commented:
Thanks for tackling this
